com.unity.perception/Runtime/Randomization/Randomizers/RandomizerTagManager.cs

```diff
@@ -9,33 +9,90 @@ namespace UnityEngine.Experimental.Perception.Randomization.Randomizers
     /// </summary>
     public class RandomizerTagManager
     {
+        Dictionary<Type, List<Type>> m_TypeCache = new Dictionary<Type, List<Type>>();
         Dictionary<Type, HashSet<GameObject>> m_TagMap = new Dictionary<Type, HashSet<GameObject>>();
+        Dictionary<GameObject, HashSet<Type>> m_ObjectTags = new Dictionary<GameObject, HashSet<Type>>();
 
         /// <summary>
         /// Enumerates all GameObjects in the scene that have a RandomizerTag of the given type
         /// </summary>
         /// <typeparam name="T">The type of RandomizerTag to query for</typeparam>
+        /// <param name="returnSubclasses">Should this method retrieve all tags derived from the passed in tag also?</param>
         /// <returns>GameObjects with the given RandomizerTag</returns>
-        public IEnumerable<GameObject> Query<T>() where T : RandomizerTag
+        public IEnumerable<GameObject> Query<T>(bool returnSubclasses = false) where T : RandomizerTag
         {
             var type = typeof(T);
+
             if (!m_TagMap.ContainsKey(type))
                 yield break;
+
             foreach (var taggedObject in m_TagMap[type])
-                yield return taggedObject;
+            {
+                if (returnSubclasses)
+                {
+                    yield return taggedObject.gameObject;
+                }
+                else
+                {
+                    if (m_ObjectTags.ContainsKey(taggedObject) && m_ObjectTags[taggedObject].Contains(type))
+                        yield return taggedObject.gameObject;
+                }
+            }
         }
 
         internal void AddTag(Type tagType, GameObject obj)
         {
-            if (!m_TagMap.ContainsKey(tagType))
-                m_TagMap[tagType] = new HashSet<GameObject>();
-            m_TagMap[tagType].Add(obj);
+            // Add tag to the game object to tag map
+            if (!m_ObjectTags.ContainsKey(obj))
+                m_ObjectTags[obj] = new HashSet<Type>();
+
+            m_ObjectTags[obj].Add(tagType);
```

Comment:
I think m_ObjectTags only ever has one tag added to any particular object's associated type hashset. Therefore, this dictionary's type could be simplified to Dictionary<GameObject, Type>